Implement NumPy-like `frombuffer` tensor constructor.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60627 Implement Python Array API `asarray` function.
* **#59077 Implement NumPy-like `from_buffer` tensor constructor.**

Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

Differential Revision: [D29870914](https://our.internmc.facebook.com/intern/diff/D29870914)

ysiraichi committed:
Implement NumPy-like `from_buffer` tensor constructor.

Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Remove unused 'unittest' import. on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

albanD reviewed (commented):
Note: I didn't look at the test
But the rest of the code looks ok, added comments inline.

Out of curiosity, why is this better than chaining np.from_buffer and torch.from_numpy?

ysiraichi commented on aten/src/ATen/templates/Functions.cpp:
```diff
@@ -105,6 +105,7 @@ Tensor TensorMaker::make_tensor() {
 
   Tensor tensor = detail::make_tensor<TensorImpl>(
       std::move(storage), opts_.computeDispatchKey(), opts_.dtype());
+  tensor.set_requires_grad(opts_.requires_grad());
```
I see. I think I can move it out to `THPVariable_from_buffer`, then. And do it after calling `for_blob`.

ysiraichi commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,109 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+      PyBuffer_Release(&view);
+
+      if (len == 0 || count == 0) {
+        PyErr_SetString(
```
The problem is that: given that I `Py_INCREF(obj)` before this check, I think I should `Py_DECREF(obj)` in case this check fails. So, I guess the ifs are needed here.

albanD replied:
If a variable has an owning reference to a PyObject, then you want to keep it in a `THPObjectPtr`. This will do the decref automatically when the object is destroyed, avoiding such leak.

You can see a few use of this in `torch/csrc/Module.cpp` in particular.

ysiraichi replied:
Hmm. I refactored the code a bit, tracking whether an error has occurred or not with an `c10::optional<std::string>` (which allows me to correctly call `Py_DECREF` accordingly). Do you think it's better to undo it, and just use `TORCH_CHECK_VALUE`?

albanD replied:
Let me check the code.

ysiraichi commented on torch/_torch_docs.py:
```diff
@@ -3375,6 +3375,53 @@ def merge_dicts(*dicts):
+
+.. warning::
+    Passing a pointer that lives in a different device than the specified
+    one results in ``RuntimeError``.
```
Yeah. It can happen, indeed. I will fix that.

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi commented:
> Out of curiosity, why is this better than chaining np.from_buffer and torch.from_numpy?

I wouldn't say it's better. I think it's more convenient, though:
- no need to use `numpy` for sharing memory
- has parameters (`requires_grad` and `device`) similar to other tensor constructors (could be implemented on from_numpy as well)
- capable of handling CUDA buffers, too (although, realistically, I'm not sure if people would use it for that)

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

albanD commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,102 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+
+    auto len = view.len;
+    auto buf = view.buf;
+    auto obj = view.obj;
```
The type of obj here should be `THPObjectPtr`.
That way you can remove all the custom logic for decref as the decref will be done automatically when the `THPObjectPtr` is destroyed.

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

albanD commented on torch/_torch_docs.py:
```diff
@@ -3385,6 +3385,55 @@ def merge_dicts(*dicts):
     array([-1,  2,  3])
 """)
 
+add_docstr(torch.from_buffer,
```
I just realised that the numpy function name doesn't have the `_`, could you update that to math the numpy naming? (https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html)

ysiraichi replied:
Indeed, it doesn't have the `_`. I left it there, to keep consistency with `from_numpy`, though. What do you think?

albanD replied:
@mruberry what do you think?

albanD commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,163 @@
+    torch.float64 : 'd',
+}
+
+def get_dtype_size(dtype):
```
nit: if you have a Tensor, you can use [element_size()](https://pytorch.org/docs/stable/tensors.html?highlight=element_size#torch.Tensor.element_size) method on Tensor to get that value.
But if you only have the dtype, this is good.

albanD commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,163 @@
+        input = self.INPUT
+        bytes = get_dtype_size(dtype)
+        # Empty array
+        with self.assertRaises(ValueError):
```
Could you use `assertRaisesRegex` and provide a distinctive part of the error message for these? Just to make sure that you're not catching an error that is not the one you expect.

albanD commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,163 @@
+    @dtypes(*all_types())
+    def test_invalid_device(self, device, dtype):
+        input = self.INPUT
+        with self.assertRaises(RuntimeError):
```
What is the error you catch here? That there are no cuda devices available?

ysiraichi replied:
Calling `TensorMaker::for_blob` saying the data pointer lives on CUDA, ends up calling `cudaPointerGetAttributes`. If we pass a pointer that was allocated on CPU, as far as I understand it, it will error out, which will be caught by `AT_CUDA_CHECK`.

ysiraichi replied:
> If we pass a pointer that was allocated on CPU, as far as I understand it, it will error out, which will be caught by `AT_CUDA_CHECK`.

Apparently, this is not true. It depends on the implementation of `cudaPointerGetAttributes` (not sure what it does when, say, it returns `cudaMemoryTypeUnregistered` as its `type`), and the context that created the pointer. Maybe it's not a good idea to have this test, then!?

albanD replied:
I feel like this test might segfault yes.
I think in this case, you just want to make the doc scary enough that users don't do it :D

albanD reviewed (commented):
Thanks for the update, the function is much simpler to follow!

I just have a couple question for the tests, the name update and this will be good to go!

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,168 @@
+
+import torch
+import array
+
```
Add a comment here about what this test class is for

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,168 @@
+class TestBufferProtocol(common.TestCase):
+    INPUT = [0, 1, 2, 3]
+
+    def _run_test(self, dtype, arr, count=-1, first=0, offset=None, **kwargs):
```
Would it be simpler to create NumPy tensors, write them to a buffer, and then compare the result of PyTorch's and NumPy's from_buffer?

ysiraichi replied:
Do you mean: instead of comparing the resulting PyTorch tensor with a slice of an array, compare it with the result of NumPy's `frombuffer`? If so, computing `last` could be avoided. But, I guess it isn't necessarily simpler.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3385,6 +3385,55 @@ def merge_dicts(*dicts):
+    before the returned tensor goes out of scope.
+
+.. warning::
+    Passing a pointer that lives in a different device than the specified
```
This seems bad. Is there some way we can address this? Alternatively, should we restrict this to CPU only for now?

mruberry replied:
Actually, I guess `asarray` might solve this with its copy kwarg? Should we be assuming that objects implementing the Python buffer protocol are on the CPU, and make a copy if `device` is a CUDA device?

cc @rgommers for this question, too

rgommers replied:
I'm not sure this warning makes sense to me. There is no raw pointer passed anywhere by the user. Maybe what is meant is that if `buffer` and `device` are mismatching, things go wrong? Better to say that then.

There's this fundamental issue here that the doc says _"a Python object that exposes the buffer interface"_, but the buffer protocol knows nothing about devices. There's only a couple of libraries that use the buffer protocol on non-CPU devices, like `mpi4py` - here's a random example I could find: https://gist.github.com/lebedov/8514d3456a94a6c73e6d.

> Actually, I guess `asarray` might solve this with its copy kwarg? Should we be assuming that objects implementing the Python buffer protocol are on the CPU, and make a copy if `device` is a CUDA device?

Never make a copy I'd say - this is a zero-copy thing. You can do it on GPU if `device` is handled correctly, it's just unfortunate that `device` has to be given by the end user which is a chance to make a mistake with the consequence of grabbing the wrong memory. That's in contrast to DLPack, where it's the library implementer's responsibility to handle devices.

Either way, let's never make a copy - just share memory or raise an exception. That's also consistent with `__array__`, `__cuda_array_interface__`, etc.

rgommers replied:
We're trying to replace existing functionality here though, right? One of the `Tensor.__init__` usages? How did that handle this issue?

ysiraichi replied:
> Maybe what is meant is that if buffer and device are mismatching, things go wrong? Better to say that then.

Yes, that's indeed the point. The way you put it is better. I will change the doc.

> There's this fundamental issue here that the doc says "a Python object that exposes the buffer interface", but the buffer protocol knows nothing about devices.

It's as you describe next. Basically, it's the user responsibility here to get the device right (which is unfortunate).

> Either way, let's never make a copy - just share memory or raise an exception.

Agreed. I also think we shouldn't copy, here (there's the tensor factory for that).

> We're trying to replace existing functionality here though, right? One of the Tensor.__init__ usages? How did that handle this issue?

The idea was to replace `new(Storage)` (construct a tensor, given a `Storage`). In _torchvision_, it was used in conjunction with `Storage.from_buffer`. This implementation:

- Assumed the buffer lived on CPU
- Copied the data to this new storage

https://github.com/pytorch/pytorch/blob/010bcb4c2dbd0bc550252ea9cac942006758fb5f/torch/csrc/generic/StorageMethods.cpp#L97

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3385,6 +3385,55 @@ def merge_dicts(*dicts):
+           r"""
+from_buffer(buffer, dtype=None, count=-1, offset=0, *, device=None, requires_grad=False) -> Tensor
+
+Create a 1-dimensional :class:`Tensor` from an object that implements
```
I was thinking about this more and wondering how the Python Array API deals with the buffer protocol. Turns out it has an `asarray` function:

https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html?highlight=frombuffer#numpy.frombuffer

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

Since our new focus is on implementing the Python Array API, do you think PyTorch would want both `asarray` and `frombuffer`, @ysiraichi, or should we just focus on building the `asarray` function? This work would be required to do it, and I'm curious to hear your thoughts.

mruberry replied:
Another option is to implement a `frombuffer` function (name TBD) that was compliant with `asarray`, so we could then build `asarray` using it

rgommers replied:
I think that's effectively why `numpy.frombuffer` is little used and doesn't quite make sense anymore - `numpy.asarray` does the same and more, and is clearer. The (PEP 3118) buffer protocol is C only and not supposed to be user-visible. Both `numpy.frombuffer` and PEP 3118 are ~15 years old by now, so I can't remember or easily find the history there anymore.

ysiraichi replied:
I think it makes sense to have both, while calling `frombuffer` in `asarray`. If we are to follow the Python Array API, I guess `frombuffer` still enables users to specify an offset and count (which is nice, but the frequency they are used probably wouldn't be high).

mruberry replied:
That sounds reasonable -- one question about that, though: is `asarray`'s buffer behavior the same as `frombuffer`'s?

ysiraichi replied:
Given that the input object is something that implements the buffer protocol, they are the same if you consider `offset=0` and `count=-1` arguments for `frombuffer`. From what I could understand, `asarray` is a function that "does the right thing", depending on what the input object is (almost a superset of `frombuffer`).

Actually, I would say that `asarray` is somewhat analogous to what we had with the legacy tensor constructor. It kind of tried to do the right thing in that context. So, in that sense, maybe we would rather have `frombuffer`, `fromdlpack`... The good thing about `asarray`, in my opinion, is that it's standardized in the Python Array API.

> `numpy.asarray` does the same and more, and is clearer.

I think `numpy.asarray` is a bit different than what the Python Array API specifies. I think it is closer to PyTorch's `as_tensor` (without the `order` parameter).

https://pytorch.org/docs/stable/generated/torch.as_tensor.html

mruberry replied:
Interesting! As long as this version of `frombuffer` is consistent from NumPy's `frombuffer` AND can be used as a helper by a future `asarray` (as defined by the Python Array API) I think it's OK.

ysiraichi replied:
I believe, as is, it's consistent with NumPy's version. It can also easily be used by a future `asarray` function, by just hoisting the code in that function into a helper one (I guess we can leave it for that PR). The only thing now is the name of the function. There are consistency reasons for both `frombuffer` and `from_buffer`, in my opinion.
- `frombuffer` is consistent with NumPy's version
- `from_buffer` is consistent with PyTorch's `Storage::from_buffer` and `torch.from_numpy`
What do you think?

mruberry replied:
Great question. I prefer `frombuffer` myself for consistency with NumPy. Sound good?

ysiraichi replied:
Yes. I will make the change!

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,85 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+    size_t actual_count = 0;
+    Py_buffer view;
+
+    if (PyObject_GetBuffer(buffer, &view, PyBUF_SIMPLE) < 0) {
```
Apparently, I missed the fact that buffers can be read-only. So, I probably should add a `PyBUF_WRITABLE` to ask for a writable there. That said, what should we do when the buffer is not writable? Should we just raise an error? Can tensors be read-only?

As is, it's possible to change read-only buffers. NumPy correctly raises an error when trying to modify the read-only buffer:

```python
> x = b"\x01\x02"
> a = torch.frombuffer(x, dtype=torch.uint8)
> a[0] = 10
> a
b'\n\x01'
```

```python
> a = numpy.frombuffer(x, dtype=torch.uint8)
> a[0] = 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: assignment destination is read-only
```

mruberry replied:
PyTorch doesn't support read-only tensors. What about being consistent with the behavior of `from_numpy()` for non-writable arrays?

```
a = np.array((1., 2., 3.))
a.flags['WRITEABLE'] = False
torch.from_numpy(a)
: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
```

So similarly we could warn the buffer is not writeable, and PyTorch does not support non-writeable tensors?

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi commented:
I think those 4 CI failures are unrelated to this PR.
@mruberry I think I've addressed all suggestions. Could you take a look at it?

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+    before the returned tensor goes out of scope.
+
+.. warning::
+    The buffer protocols has no notion of devices. Thus, this function only
```
How could a CUDA buffer occur? Is there any way to detect this case or protect ourselves if it happens? Segfaults are bad.

ysiraichi replied:
As far as I know, there is [`pycuda`](https://documen.tician.de/pycuda/tutorial.html) that provides CUDA buffers, and `mpi4py` that can use CUDA buffers ([link @rgommers referenced](https://gist.github.com/lebedov/8514d3456a94a6c73e6d)).

I am not sure how we could detect CUDA buffers, though. That is because, it all boils down to being able to tell CUDA and CPU pointers apart. AFAIK, it's something hard to be done, since a given pointer can be valid on both of them...

mruberry replied:
Thanks for the explanation. I guess we'll stick with this for now

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> import array
```
Nice example

mruberry reviewed (commented):
This is looking really good, @ysiraichi. I still want to understand the CUDA buffer concern better. How could they occur, and can we protect ourselves from them if they do?

@rgommers would you like to take a look at this, too?

rgommers commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,184 @@
+    torch.int32   : 'i',
+    torch.int64   : 'q',
+    torch.float32 : 'f',
+    torch.float64 : 'd',
```
Maybe I have a bad memory but: why no `complex128`, `complex256`? Worth a comment if it's left out on purpose.

ysiraichi replied:
Because [the python array module](https://docs.python.org/3/library/array.html) doesn't have typecodes for those.

rgommers replied:
Ah, but this is the buffer protocol - which does have them: https://docs.python.org/3/c-api/buffer.html#complex-arrays. The buffer protocol docs are terrible, and don't give you these type codes. Maybe easiest to look at the source code for it, or for how NumPy processes the buffer protocol.

ysiraichi replied:
Huh. Maybe I'm missing something, but I can't find where it describes the complex types (as far as I understood it, it describes how to deal with arrays with strides and suboffsets).

Anyways, this table is used for converting PyTorch dtypes into Python Array module's typecodes (and, apparently, it doesn't support complex types). But, maybe, instead of the array module, I could be using NumPy (since it does implement the buffer protocol) as [suggested by @mruberry previously](https://github.com/pytorch/pytorch/pull/59077#discussion_r646091867).

rgommers commented:
There's a couple of unanswered/unresolved comments left further up in the review (click `Load more ...`). 

> @rgommers would you like to take a look at this, too?

This looks quite good to me!

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi commented:
@albanD @mruberry @rgommers 
Thank you for your review. I think, now, I have addressed all the comments.
I have also replaced the `array` module with NumPy (now, we test for all dtypes, at least, those exchangeable with NumPy) and stopped using `torch.iinfo` and `torch.finfo` (using `torch.Tensor.element_size` as mentioned earlier).
As for the 9 CI failures, they appear to come from the base commit, so not related to this PR.

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,99 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    "frombuffer(PyObject* buffer, ScalarType dtype=None, int64_t count=-1, int64_t offset=0, *, bool requires_grad=False)",
```
What would you think of making everything after "buffer" kwarg-only? This would be more consistent with other PyTorch operations that take "tensor options" (like dtype, device, requires_grad...) and the Python Array API's style

ysiraichi replied:
I think that is probably a good idea. Another thing that would probably be better is to make `dtype` a required argument. What do you think? Since we are blindly believing in the specified `dtype`, maybe we could make the user explicitly specify it so as not to have any (less) surprising behaviors.

mruberry replied:
Making it required for now seems OK

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,99 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+      TORCH_CHECK(
+          PyObject_GetBuffer(buffer, &view, PyBUF_SIMPLE) >= 0,
+          "could not retrieve buffer from object");
+      TORCH_WARN_ONCE(
```
Nice use of WARN_ONCE

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,99 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+    auto buf = view.buf;
+    PyBuffer_Release(&view);
+
+    TORCH_CHECK_VALUE(
```
Nice error check

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,99 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+                        Py_DECREF(obj);
+                      })
+                      .make_tensor();
+    tensor.requires_grad_(requires_grad);
```
Nit: can tensoroptions specify requires grad?

ysiraichi replied:
Yes, it can! I completely missed that.

ysiraichi replied:
In fact, `make_tensor` doesn't use `requires_grad`. So, we have to set it outside.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.frombuffer,
+           r"""
+frombuffer(buffer, dtype=None, count=-1, offset=0, *, requires_grad=False) -> Tensor
```
If you like the kwarg idea, this signature would become 

```
buffer, *, ...
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+Create a 1-dimensional :class:`Tensor` from an object that implements
+the Python buffer protocol.
+
+Interpret :attr:`buffer` as a 1-dimensional buffer of type :attr:`dtype`
```
Nit: "Create" -> "Creates" on line 3594 and "Interpret" -> "Interprets" here. 

What happens if the dtype is not specified? Maybe this paragraph should add a sentence at the end which explains the default behavior: "By default, the returned tensor has dtype X and contains all the values in the buffer."

mruberry replied:
Actually, follow-up question: is that the default? Is the buffer always interpreted as a byte buffer, or are the values in the buffer interpreted as values? Like if I have a Python array [1, 2, 3] and I interpret that in float, do I get back [1., 2., 3.]?

mruberry replied:
It looks like NumPy always interprets the array as bytes and just writes to the bytes directly

ysiraichi replied:
> What happens if the dtype is not specified? Maybe this paragraph should add a sentence at the end which explains the default behavior: "By default, the returned tensor has dtype X and contains all the values in the buffer."

When `dtype` is not specified, the default scalar type is used. I will add it to the docs.

> is that the default? Is the buffer always interpreted as a byte buffer, or are the values in the buffer interpreted as values?

This implementation behaves the same as NumPy. So we use the bytes in the buffer as they are, but interpret them as the specified `dtype`. Maybe I should change the wording there, since it's not completely clear. What do you think about:

_"Skip the first `offset` bytes in `buffer`, and interpret the rest of them as a 1-dimensional tensor of type `dtype` with `count` elements"_

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -465,6 +467,99 @@ static PyObject * THPVariable_get_device(PyObject* self_, PyObject* args, PyObje
+    auto offset = r.toInt64(3);
+    auto requires_grad = r.toBool(4);
+
+    auto elsize = at::elementSize(dtype);
```
How does this work if dtype is None?

ysiraichi replied:
Calling `r.scalartype(..)` returns the default `ScalarType` (i.e. `torch::tensors::get_default_scalar_type()`).

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+Interpret :attr:`buffer` as a 1-dimensional buffer of type :attr:`dtype`
+with :attr:`count` elements after the first :attr:`offset` bytes.
+
+The returned tensor and buffer share the same memory. Modifications to
```
How does this work if the buffer has a different dtype then what was specified?

mruberry replied:
NumPy seems to interpret the buffer as byte data and writes directly to the bytes, which ay be surprising. It'd be nice to clarify this in the documentation if that's the approach we want to take.

ysiraichi replied:
It has the same behavior as NumPy. It blindly believes the specified `dtype`, and interprets the data accordingly.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+
+.. warning::
+    The buffer protocols has no notion of devices. Thus, this function only
+    works on CPU buffers. Passing CUDA buffers will probably lead to
```
What about this:

"This function's behavior is undefined when passed an object implementing the buffer protocol whose data is not on the CPU. Doing so is likely to cause a segmentation fault." 

?

ysiraichi replied:
That's way clearer! Thanks for the suggestion.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+    segmentation fault (at best), but could also silently yield wrong results.
+
+Args:
+    buffer (object): a Python object that exposes the buffer interface
```
Nit: period at end of this sentence

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+        the buffer. Default: 0.
+
+Keyword args:
+    {requires_grad}
```
If more arguments are made kwarg-only they need to be moved into this section

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3587,6 +3587,54 @@ def merge_dicts(*dicts):
+    tensor([ 1,  2,  3])
+    >>> t[0] = -1
+    >>> a
+    array([-1,  2,  3])
```
Should we add an example with a comment of creating a tensor with a different dtype than the array?

ysiraichi replied:
That sounds like a good idea.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+                                       r"The given buffer is not writable."):
+            torch.frombuffer(byte_arr, dtype=dtype)
+
+instantiate_device_type_tests(TestBufferProtocol, globals())
```
`instantiate_device_type_tests(TestBufferProtocol, globals(), only_for='cpu')` will only create tests for  the CPU so the `@onlyCPU` decorator won't be needed on each test

ysiraichi replied:
That's definitely way better! Thanks for the tip.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+import numpy
+
+def get_dtype_size(dtype):
+    return int(torch.empty((), dtype=dtype).element_size())
```
Not for this PR, but I wonder if we should let users acquire a dtypes info directly from it. So, like, `torch.float32.info` would be equivalent to `torch.finfo(torch.float32)`.

ysiraichi replied:
I feel like that would probably be a good thing. Not sure about how frequently it shows up on users' code, though. By the way, do complex values have anything similar: `torch.cinfo` (didn't find it)?

mruberry replied:
A natural question. The complex types are accessed using finfo, actually. This is consistent with NumPy.

ysiraichi replied:
Ah, I guess I missed the fact that it was possible to pass an argument to `torch.finfo` constructor.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+SIZE = 5
+SHAPE = (SIZE,)
+
+# Tests for the `frombuffer` function (only work on CPU):
```
Nice comment

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+        return (numpy_original, torch_frombuffer)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
```
Nice dtype enumeration

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_same_type(self, device, dtype):
+        self._run_test((), dtype)
```
Good test shapes

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+            _, tensor = self._run_test(*args, **kwargs)
+            self.assertTrue(tensor.requires_grad == requires_grad)
+
+        if dtype.is_floating_point or dtype.is_complex:
```
Nit: this could also be modeled like this:

```
requires_grad = dtype.is_floating_point or dtype.is_complex
_run_test_and_check_grad(requires_grad, (), dtype)
_run_test_and_check_grad(requires_grad, (4,), dtype)
_run_test_and_check_grad(requires_grad, (10, 10), dtype)
```

ysiraichi replied:
Ah, very nice observation. Thanks.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_with_offset(self, device, dtype):
+        # Offset should be valid whenever there is, at least,
```
This is true when the dtype passed to frombuffer is the same as the data's dtype, but not true generally. 

Are tests when the dtypes differ interesting?

ysiraichi replied:
> Are tests where the dtypes differ interesting?

Maybe for converting from bytes loaded at some point to 4-byte integers. I will add one test for that.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_with_count_and_offset(self, device, dtype):
+        # Explicit default count [-1, 1, 2, ..., len]
+        for i in range(-1, SIZE + 1):
```
This is fine and doesn't need to change, but in the future it might be easier to just handwrite some sample inputs for offset, count, and their combination, instead of exhaustively enumerating all possible combinations

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_invalid_positional_args(self, device, dtype):
```
Great test for invalid inputs and error messages

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+                                        rf"offset \({offset} bytes\) must be"):
+                self._run_test(SHAPE, dtype, offset=bytes - 1)
+        # Count too big for each good first element
+        for first in range(SIZE):
```
This is fine and doesn't need to change in this PR, but in the future it might be simpler to just try one invalid input of this type instead of enumerating every invalid input

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+                continue
+
+            actual_count = count if count > 0 else SIZE
+            for first in range(SIZE - actual_count):
```
This is OK but it would probably also be fine to just manipulate a few values.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+            for first in range(SIZE - actual_count):
+                last = first + actual_count
+                arr, tensor = self._run_test(SHAPE, dtype, first=first, count=count)
+                tensor[:] = x
```
What about editing a value in the array and ensuring that appears properly in the tensor?

ysiraichi replied:
That makes sense. I will add it.

mruberry commented on test/test_buffer_protocol.py:
```diff
@@ -0,0 +1,175 @@
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_not_a_buffer(self, device, dtype):
```
Nice test for non-buffers

mruberry reviewed (commented):
Hey @ysiraichi! 

Overall this is really good, and I think it's close to be doing done. I made some comments inline for your review. In particular, I suggest we think about revising the docs just a little bit to explain that the buffer is interpreted as a byte tensor, and what happens if the dtype of the tensor and buffer's type differ. I also made a few suggestions to extend test coverage. 

Many suggestions are just "nits" that are small comments that don't need to be addressed for this PR to be merged, so don't worry about those.

Looking forward to getting this merged!

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

mruberry reviewed (commented):
Hey @ysiraichi! Just checking in. I read through the current version of the PR and think it looks good. What are your thoughts? Is it ready to merge?

ysiraichi committed:
Update on "Implement NumPy-like `from_buffer` tensor constructor."


Fixes #58549

`from_buffer` constructs a tensor object from an already allocated buffer through
CPython's buffer protocol. Besides the standard `dtype`, `count`, and `offset` parameters,
this function also accepts:

- `device`: where the buffer lives
- `requires_grad`: should autograd record operations on the new tensor

A new test file _test_buffer_protocol.py_ was created. Currently, only CPU tests were
implemented. That's because neither PyTorch nor Numba implements CPython's buffer
protocol. Therefore, there's no way to create a CUDA buffer with the existing
dependencies (could use PyCUDA for that, though).

At the moment, if `device` differs from the device the buffer actually lives, two things
may happen:

- `RuntimeError`, if `device='cuda'`
- Segmentation fault (not tested -- see above), if `device='cpu'`

[ghstack-poisoned]

ysiraichi commented:
@mruberry I think so. I just had to add one last change in `gen_pyi.py`: make `dtype` required by changing its type to `_dtype`.

mruberry reviewed (approved):
LGTM, @ysiraichi! Thanks for implementing frombuffer!

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29870914).
